File: aoc/2019/aoc201925.go
package main

import (
	"io"
	"os"
)

func part1(canned []byte, script io.Reader, ic *intcode) {
	buf := [20]byte{}
	inp := canned
	showOutput := false
	ic.init()
	ic.interpIO(func() (bool, int64) {
		if len(inp) > 0 {
			ch := inp[0]
			inp = inp[1:]
			return false, int64(ch)
		}
		showOutput = true
		trace = false
		for i := 0; i < 20; i++ {
			script.Read(buf[i:i+1])
			if buf[i] == '\n' {
				if buf[0] == '@' {
					trace = true
					inp = buf[2:i+1]
					return false, int64(buf[1])
				}
				inp = buf[1:i+1]
				return false, int64(buf[0])
			}
		}
		inp = buf[1:]
		return false, int64(buf[0])
	}, func(out int64) bool {
		if showOutput {
			print(string([]byte{byte(out)}))
		}
		return false
	})
}

func main() {
	const canned = `east
south
south
take hologram
north
north
west
south
take mouse
east
take shell
west
west
take whirled peas
east
north
west
north
north
west
take semiconductor
east
south
west
south
take hypercube
north
east
south
west
take antenna
south
take spool of cat6
north
west
south
`
	part1([]byte(canned), os.Stdin, parseFile("input/25.txt"))
}

// For my input data:
// hypercube:     0x       1
// shell:         0x  800000
// whirled peas:  0x      80
// spool of cat6: 0x10000000
// mouse:         0x       2
// antenna:       0x    4000
// hologram:      0x 1000000
// semiconductor: 0x    1000

File: aoc/2024/aoc201925.hs
module AOC201925 where

import Data.Char(chr,ord)
import Data.Vector.Unboxed(Vector)

import AOC
import AOC2019

aoc = AOC {
    day="../../2019/input/25",
    aocTests=[],
    aocCode=Code {
        codeParse=parseIntCode,
        codeParse2=parseIntCode,
        codeTest=undefined,
        codeTest2=undefined,
        codeResult=result,
        codeResult2=return ()
        }
    }

interactIntCode :: (state -> String -> (state,String)) -> state -> String
                -> Vector Int -> String
interactIntCode makeInput initialState initialInput =
    intCodeIO (IntCodeIO (continue initialState [])) (map ord initialInput)
  where
    continue state outBuffer (Just output) input Nothing =
        map chr (reverse (output:outBuffer))
    continue state outBuffer (Just output) input (Just cont) =
        cont (IntCodeIO (continue state (output:outBuffer))) input
    continue state outBuffer Nothing input@(_:_) (Just cont) =
        cont (IntCodeIO (continue state outBuffer)) input
    continue state outBuffer Nothing [] (Just cont) =
        cont (IntCodeIO (continue newState [])) (map ord newInput)
      where
        (newState,newInput) = makeInput state (map chr (reverse outBuffer))

findWeight (item:items) text
  | elem "heavier" $ words text =
      (items,unlines ["take " ++ item, "drop " ++ head items, "south"])
  | null items = (items,"south\n")
  | otherwise = (items,unlines ["drop " ++ head items, "south"])

result = extract . interactIntCode findWeight items script
  where
    extract = head . drop 1 . dropWhile (/= "typing") . words

    script = unlines [
        "east",
        "south",
        "south",
        "take hologram",
        "north",
        "north",
        "west",
        "south",
        "take mouse",
        "east",
        "take shell",
        "west",
        "west",
        "take whirled peas",
        "east",
        "north",
        "west",
        "north",
        "north",
        "west",
        "take semiconductor",
        "east",
        "south",
        "west",
        "south",
        "take hypercube",
        "north",
        "east",
        "south",
        "west",
        "take antenna",
        "south",
        "take spool of cat6",
        "north",
        "west",
        "south",
        "drop spool of cat6"
        ]

    items = [
        "spool of cat6",
        "hologram",
        "shell",
        "antenna",
        "semiconductor",
        "whirled peas",
        "mouse",
        "hypercube"
        ]
